pkg/skaffold/runner/runner.go

```diff
@@ -139,14 +154,79 @@ func NewForConfig(opts *config.SkaffoldOptions, cfg *latest.SkaffoldConfig) (*Sk
 			Monitor: monitor,
 			Trigger: trigger,
 		},
-		changeSet:            &changeSet{},
+		changeSet: &changeSet{
+			rebuildTracker: make(map[string]*latest.Artifact),
+			resyncTracker:  make(map[string]*sync.Item),
+		},
 		labellers:            labellers,
 		defaultLabeller:      defaultLabeller,
 		portForwardResources: cfg.PortForward,
 		imageList:            kubernetes.NewImageList(),
 		cache:                artifactCache,
 		runCtx:               runCtx,
-	}, nil
+		intentLock:           &gosync.Mutex{},
+		resetBuildIntent:     func() {},
+		resetSyncIntent:      func() {},
+		resetDeployIntent:    func() {},
+	}
+
+	r.setupTriggerCallbacks()
+
+	return r, nil
+}
+
+func (r *SkaffoldRunner) setupTriggerCallbacks() {
+	r.buildIntent = true
+	api := "api"
+	if strings.ToLower(r.runCtx.Opts.BuildTrigger) == api || r.runCtx.Opts.APIMode {
+		r.buildIntent = false
+		buildTriggerCallback := func() {
+			logrus.Info("build intent received, calling back to runner...")
+			r.intentLock.Lock()
```

Comment:
I would try to move the locking logic into the `intents` type

Reply:
done, lmk what you think